manifest/list.go

```diff
@@ -0,0 +1,119 @@
+	ToOCI1Index() (*OCI1Index, error)
+
+	// ToSchema2List() returns the list rebuilt as a Schema2 list, converting it if necessary.
+	ToSchema2List() (*Schema2List, error)
+}
+
+// ManifestListUpdate includes the fields which a manifest's UpdateInstances() method will modify.
+type ManifestListUpdate struct {
+	Digest    digest.Digest
+	Size      int64
+	MediaType string
+}
+
+// dupStringSlice returns a deep copy of a slice of strings, or nil if the
+// source slice is empty.
+func dupStringSlice(list []string) []string {
+	if len(list) == 0 {
+		return nil
+	}
+	dup := make([]string, len(list))
+	for i := range list {
+		dup[i] = list[i]
+	}
+	return dup
+}
+
+// dupStringStringMape returns a deep copy of a map[string]string, always returning
+// an initialized map, even if there are no contents to copy into it.
+func dupStringStringMap(m map[string]string) map[string]string {
+	if len(m) == 0 {
+		return nil
+	}
+	result := make(map[string]string)
+	for k, v := range m {
+		result[k] = v
+	}
+	return result
+}
+
+// ListFromBlob parses a list of manifests.
+func ListFromBlob(manifest []byte, manifestMIMEType string) (ManifestList, error) {
+	switch normalized := NormalizedMIMEType(manifestMIMEType); normalized {
+	case DockerV2ListMediaType:
+		return Schema2ListFromManifest(manifest)
+	case imgspecv1.MediaTypeImageIndex:
+		return OCI1IndexFromManifest(manifest)
+	case DockerV2Schema1MediaType, DockerV2Schema1SignedMediaType:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	case imgspecv1.MediaTypeImageManifest:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	case DockerV2Schema2MediaType:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	default:
+		return nil, fmt.Errorf("Unimplemented manifest MIME type %s (normalized as %s)", manifestMIMEType, normalized)
+	}
+	return nil, fmt.Errorf("Unimplemented manifest MIME type %s", manifestMIMEType)
+}
+
+// ChooseManifestInstanceFromManifestList returns a digest of a manifest appropriate
+// for the current system from the manifest available from src.
+func ChooseManifestInstanceFromManifestList(ctx *types.SystemContext, manifests ManifestList) (digest.Digest, error) {
+	mt := manifests.MIMEType()
+	if mt != DockerV2ListMediaType && mt != imgspecv1.MediaTypeImageIndex {
+		return "", fmt.Errorf("Internal error: Trying to select an image from a non-manifest-list manifest type %s", mt)
+	}
+	return manifests.ChooseDigest(ctx)
+}
+
+// computeListID computes an image ID using the list of images referred to in a ManifestList.
+func computeListID(manifests ManifestList) string {
+	instances := manifests.Instances()
+	hexes := make([]string, len(instances))
+	for i, manifest := range manifests.Instances() {
+		hexes[i] = manifest.Digest.Hex()
+	}
+	sort.Strings(hexes)
+	return digest.FromBytes([]byte(strings.Join(hexes, ""))).Hex()
+}
```

Comment:
This is a guesswork implementation.  If you have a better idea or know of a reference for a better way to compute a stable value, we should replace this.

Reply:
Right, that information doesn't seem to live in the image structure, either, which would have been handy.  Changed it to hopefully work more like what you describe.